### PR TITLE
expression, util: handle err instead of panic in EncodeDecimal

### PR DESCRIPTION
**Pull request by XuHuaiyu**
PTAL @zz-jason @coocood 
Thank you for working on TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

## What have you changed? (mandatory)
Before this commit, we `panic` when meets error in EncodeDecimal.
This commit handle the error instead of panic.

## What are the type of the changes (mandatory)?
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

unit test

## Does this PR affect documentation (docs/docs-cn) update? (optional)
no

## Refer to a related PR or issue link (optional)
split from #6658

### CODE REVIEW

**Comment by jackysp** [comment]
Could you add an integration test for this pr?

File: util/codec/codec.go [hunk 1]
```diff
@@ -97,7 +97,12 @@ func encode(sc *stmtctx.StatementContext, b []byte, vals []types.Datum, comparab
 				}
 				b = append(b, bin...)
 			} else {
-				b = EncodeDecimal(b, vals[i].GetMysqlDecimal(), vals[i].Length(), vals[i].Frac())
+				b, err = EncodeDecimal(b, vals[i].GetMysqlDecimal(), vals[i].Length(), vals[i].Frac())
+				if terror.ErrorEqual(err, types.ErrTruncated) {
+					err = sc.HandleTruncate(err)
```
Comment (jackysp) [inline util/codec/codec.go]: Do we need a test case to cover this?

**Comment by XuHuaiyu** [comment]
@jackysp I can not find an integration test to cover that,
but it'll cause some unexpected error when supporting this feature #6658.

**Comment by coocood** [comment]
LGTM

**Pull request merged**